arroyo/processing/strategies/run_task_with_multiprocessing.py

```diff
@@ -601,23 +610,14 @@ def join(self, timeout: Optional[float] = None) -> None:
 
         logger.debug("Waiting for %s batches...", len(self.__processes))
 
-        while self.__processes:
-            try:
-                self.__check_for_results(
-                    timeout=max(deadline - time.time(), 0)
-                    if deadline is not None
-                    else None
-                )
-            except multiprocessing.TimeoutError:
-                pass
+        self.__check_for_results(
+            timeout=timeout,
+        )
 
         self.__pool.close()
 
         logger.debug("Waiting for %s...", self.__pool)
-        # ``Pool.join`` doesn't accept a timeout (?!) but this really shouldn't
-        # block for any significant amount of time unless something really went
-        # wrong (i.e. we lost track of a task)
-        self.__pool.join()
+        self.__pool.terminate()
```

Comment:
Is this change actually needed now that RunTaskWithMultiprocessing.join() is applying the proper timeout in the check for results step? I think it's better to shut down cleanly with `pool.close()` + `pool.join()` rather than `terminate()`.

Reply:
if you do that and submit a task that takes 10 seconds, and call `join(timeout=3)`, `__check_for_results` will return after 3 seconds. then, `pool.join()` will block for another 7 seconds because the task is still not done

Reply:
Ok. Is there a reason for keeping the `pool.close()` call directly before this?

Reply:
Ah, no. I also wonder if we should block on the pool shutting down by calling terminate + join instead of just terminate. I am not sure we should really block rebalance on it

Reply:
`terminate()` makes sense to me if we have exceeded the deadline already. Though if no timeout is passed, I think it's more correct to honor that and close + join instead of terminate.

Reply:
by the time we call `terminate()`, the deadline is either guaranteed to have passed, or the pool is guaranteed to be empty.

If no deadline is set, `__check_for_results` will block until the pool is empty. So if we `join` after that, it can only either block for (much) longer than the deadline allows, or do nothing useful at all

Reply:
Ah I see, terminate makes sense then. Still, why do we call `pool.close()` before `pool.terminate()`?

Reply:
Oh yeah I forgot about that, will fix on Monday